oauth2_http/java/com/google/auth/oauth2/TrustBoundary.java

```diff
@@ -0,0 +1,192 @@
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.base.MoreObjects;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Represents a trust boundary that can be used to restrict access to resources. This is an
+ * experimental feature.
+ */
+public final class TrustBoundary {
+
+  static final String TRUST_BOUNDARY_KEY = "x-allowed-locations";
+  static final String GOOGLE_AUTH_TRUST_BOUNDARY_ENABLED_ENV_VAR =
+      "GOOGLE_AUTH_TRUST_BOUNDARY_ENABLE_EXPERIMENT";
+  private static final String NO_OP_VALUE = "0x0";
+  private final String encodedLocations;
+  private final List<String> locations;
+
+  public TrustBoundary(String encodedLocations, List<String> locations) {
+    this.encodedLocations = encodedLocations;
+    this.locations =
+        locations == null
+            ? Collections.<String>emptyList()
+            : Collections.unmodifiableList(locations);
+  }
+
+  private static EnvironmentProvider environmentProvider = SystemEnvironmentProvider.getInstance();
+
+  public String getEncodedLocations() {
+    return encodedLocations;
+  }
+
+  public List<String> getLocations() {
+    return locations;
+  }
+
+  public boolean isNoOp() {
+    return NO_OP_VALUE.equals(encodedLocations);
+  }
+
+  /** Represents the JSON response from the trust boundary endpoint. */
+  public static class TrustBoundaryResponse extends GenericJson {
+    @Key("encodedLocations")
+    private String encodedLocations;
+
+    @Key("locations")
+    private List<String> locations;
+
+    public String getEncodedLocations() {
+      return encodedLocations;
+    }
+
+    public List<String> getLocations() {
+      return locations;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("encodedLocations", encodedLocations)
+          .add("locations", locations)
+          .toString();
+    }
+  }
+
+  // expose this setter only for testing purposes
+  public static void setEnvironmentProviderForTest(@Nullable EnvironmentProvider provider) {
+    environmentProvider = provider == null ? SystemEnvironmentProvider.getInstance() : provider;
+  }
+
+  static boolean isTrustBoundaryEnabled() throws IOException {
+    String tbEnabled = environmentProvider.getEnv(GOOGLE_AUTH_TRUST_BOUNDARY_ENABLED_ENV_VAR);
+    if (tbEnabled == null) {
+      return false;
+    }
+    String lowercasedTbEnabled = tbEnabled.toLowerCase();
+    if ("true".equals(lowercasedTbEnabled) || "1".equals(tbEnabled)) {
+      return true;
+    }
+    if ("false".equals(lowercasedTbEnabled) || "0".equals(tbEnabled)) {
+      return false;
+    }
+    throw new IOException(
+        String.format(
+            "Invalid value for %s environment variable: \"%s\". Supported values are 'true', '1', 'false', or '0'.",
+            GOOGLE_AUTH_TRUST_BOUNDARY_ENABLED_ENV_VAR, tbEnabled));
```

Comment:
IIUC, I think this should just be `return false`? The feature gate would be any value other than "true" or "1"?